File: leetcode/medium/974leetcode.java
//974. Subarray Sums Divisible by K

//O(n*n) TLE

class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int count = 0;
        for(int i = 0; i < nums.length; i++){
            int sum = 0;
            for(int j = i; j < nums.length; j++){
                sum += nums[j];
                if(sum % k == 0) count++;
            }
        }
        return count;
    }
}


//O(n)
class Solution {
    //map(rem, frequecy)
    public int subarraysDivByK(int[] nums, int k) {
        int count = 0;

        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0,1);

        int sum = 0, rem = 0;
        for(int i = 0; i < nums.length; i++){
            sum += nums[i];
            rem = sum%k;
            if(rem < 0) rem+= k;

            if(map.containsKey(rem)){
                count += map.get(rem);
                map.put(rem, map.get(rem) + 1);
            }else{
                map.put(rem, 1);
            }

        }
        return count;
    }
}

File: leetcode/medium/974leetcode.cpp
Subarray Sums Divisible by K
Given an integer array nums and an integer k, return the number of non-empty subarrays that have a sum divisible by k.

A subarray is a contiguous part of an array.
  
  class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int n=nums.size();
        unordered_map<int,int>mp;
        int cnt=0,sum=0;
        mp[0]=1;
        for(int i=0;i<n;i++){
            sum+=nums[i];
            int rem=sum%k;
            if(rem<0)
                rem+=k;
            if(mp.count(rem)!=0)
                cnt+=mp[rem];
            mp[rem]++;
        }
        return cnt;
    }
};
